plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorNaming.xtend

```diff
@@ -69,8 +74,25 @@ class XtextGeneratorNaming {
 		new TypeReference(grammar.eclipsePluginBasePackage, getSimpleName(grammar) + 'ExecutableExtensionFactory')
 	}
 	
-	def getEclipsePluginActivator(Grammar grammar) {
-		return new TypeReference(grammar.eclipsePluginBasePackage + '.internal',getSimpleName(grammar) + 'Activator')
+	def getEclipsePluginActivator() {
+		val pluginName = projectConfig.eclipsePlugin.name
+		
+		// start determining the activator's name by stripping the common ".ui" suffix
+		var activatorName = pluginName.replaceAll('\\.ui$', '')
+		
+		// build the activator's name based on the last segment of the stripped FQN
+		//  the call of 'new TypeReference(...)' below checks for invalid characters
+		activatorName = activatorName.substring(activatorName.lastIndexOf('.') + 1).toFirstUpper
+		
+		if (activatorName != pluginName) {
+			// ... pluginName ends with ".ui" as by default, so
+			activatorName += 'UIActivator'
```

Comment:
To indicate that it's not the language defining plug-in's activator but that of the plug-in contributing the UI stuff.
But it's not that important.